recipes-app/iot2050-firmware-update/files/iot2050-firmware-update

```diff
@@ -312,15 +340,66 @@ class ManagedFirmwareUpdate(FirmwareUpdate):
             while choice > len(firmware_names) or choice < 1:
                 print("Out of range, please reinput your choice:")
                 choice = int(input("-> "))
-            
+
             firmware_name = firmware_names[choice - 1]
 
         print("Will write {} to the SPI flash".format(firmware_name))
 
         self.firmware.seek(0)
-        self.firmware = tarfile.open(fileobj=self.firmware).extractfile(firmware_name)
-        super().update_firmware()
+        with tarfile.open(fileobj=self.firmware) as f:
+            self.firmware = f.extractfile(firmware_name)
+            f.extract(member=self.uboot_default_env_file, path='.')
+            super().update_firmware()
+
+    def update_uboot_env(self, env_list):
+        '''
+        restore uboot env
+        retrive the uboot default env file belong to the upgrade firmware.
+        makeup the restore env binary and flash to the env mtd partition
+        '''
+        env_default_binary = 'uEnv_default.bin'
```

Comment:
This does not always work if the current dir is read-only.

Maybe try to use some dirs that alreadys has the write permission, such as `/tmp/`.

Also, if `mkenvimage` could have the output piped, you can redirect the output to the std_out and read it via subprocess, IIRC, then the local file could be avoided.

Reply:
I am not sure `mkenvimage` has the output piped. maybe the output file is necessary.

Reply:
Just verified via 

```
mkenvimage -s 1024 -r input.txt | hexdump
```

It works.

Reply:
new version use  `/tmp` folder, when reboot, these tmp file would not exist